### PR TITLE
Add documentation for remote cluster proxy mode

### PR DESCRIPTION
**Pull request by Tim-Brooks**
This is related to #49067.

### CODE REVIEW

**Comment by tbrooks8** [comment]
Depends on #52796

File: docs/reference/modules/remote-clusters.asciidoc [hunk 1]
```diff
@@ -15,21 +15,32 @@ endif::[]
...
 // tag::how-remote-clusters-work[]
 Remote cluster connections work by configuring a remote cluster and connecting
-only to a limited number of nodes in that remote cluster. Each remote cluster
-is referenced by a name and a list of seed nodes. When a remote cluster is
-registered, its cluster state is retrieved from one of the seed nodes and up
-to three _gateway nodes_ are selected to be connected to as part of remote
-cluster requests.
-// end::how-remote-clusters-work[]
+only to a limited number of nodes in that remote cluster. There are two
+potential modes for remote cluster connections: sniff mode and proxy mode.
```
Comment (ywelsch) [inline docs/reference/modules/remote-clusters.asciidoc]: I would remove the word "potential" here.

Comment (ywelsch) [reply docs/reference/modules/remote-clusters.asciidoc]: also, perhaps link sniff and proxy to the corresponding sections below?

File: docs/reference/modules/remote-clusters.asciidoc [hunk 2]
```diff
@@ -201,23 +246,15 @@ PUT _cluster/settings
 [[remote-cluster-settings]]
 === Remote cluster settings
 
-`cluster.remote.connections_per_cluster`::
-
-  The number of gateway nodes to connect to per remote cluster. The default is
-  `3`.
+`cluster.remote.${cluster_alias}.mode`::
```
Comment (ywelsch) [inline docs/reference/modules/remote-clusters.asciidoc]: Perhaps add a sentence above the first setting stating that this section is covering settings that apply to both sniff and proxy mode, and that mode-specific settings are coming later.

File: docs/reference/modules/remote-clusters.asciidoc [hunk 3]
```diff
@@ -201,23 +246,15 @@ PUT _cluster/settings
 [[remote-cluster-settings]]
 === Remote cluster settings
```
Comment (ywelsch) [inline docs/reference/modules/remote-clusters.asciidoc]: Remote cluster settings for all modes?

File: docs/reference/modules/remote-clusters.asciidoc [hunk 4]
```diff
@@ -251,6 +288,49 @@ PUT _cluster/settings
   Elasticsearch compresses the response. If unset, the global
   `transport.compress` is used as the fallback setting.
 
+[float]
+[[remote-cluster-sniff-settings]]
+=== Remote cluster sniff mode settings
```
Comment (ywelsch) [inline docs/reference/modules/remote-clusters.asciidoc]: Remote cluster settings for sniff mode?

**Review by ywelsch — COMMENTED** [review]
@jrodewig can you have a look too?

File: docs/reference/cluster/remote-info.asciidoc [hunk 5]
```diff
@@ -17,29 +17,47 @@ Returns configured remote cluster information.
...
 by the configured remote cluster alias.
 
 
 [[cluster-remote-info-api-response-body]]
 ==== {api-response-body-title}
 
-`seeds`::
-	The configured initial seed transport addresses of the remote cluster.
+`mode`::
+    The configured connection mode for the remote cluster. Will be `sniff` or
```
Comment (jrodewig) [inline docs/reference/cluster/remote-info.asciidoc]:
```suggestion
    Configured connection mode for the remote cluster. Returned values are `sniff` or
```

File: docs/reference/cluster/remote-info.asciidoc [hunk 6]
```diff
@@ -17,29 +17,47 @@ Returns configured remote cluster information.
...
 	The initial connect timeout for remote cluster connections.
 
 [[skip-unavailable]]
 `skip_unavailable`::
     Whether the remote cluster is skipped in case it is searched through
     a {ccs} request but none of its nodes are available.
+
+`seeds`::
+	The configured initial seed transport addresses of the remote cluster when
+	sniff mode is configured.
```
Comment (jrodewig) [inline docs/reference/cluster/remote-info.asciidoc]: This looks like it was double indented.

```suggestion
`seeds`::
    Initial seed transport addresses of the remote cluster when
    `sniff` mode is configured.
```

File: docs/reference/cluster/remote-info.asciidoc [hunk 7]
```diff
@@ -17,29 +17,47 @@ Returns configured remote cluster information.
...
+
+`num_nodes_connected`::
+    The number of connected nodes in the remote cluster when sniff mode is
+    configured.
```
Comment (jrodewig) [inline docs/reference/cluster/remote-info.asciidoc]:
```suggestion
    Number of connected nodes in the remote cluster when `sniff` mode is
    configured.
```

File: docs/reference/cluster/remote-info.asciidoc [hunk 8]
```diff
@@ -17,29 +17,47 @@ Returns configured remote cluster information.
...
+
+`max_connections_per_cluster`::
+	The maximum number of connections maintained for the remote cluster when
+	sniff mode is configured.
```
Comment (jrodewig) [inline docs/reference/cluster/remote-info.asciidoc]: Double indented.
```suggestion
`max_connections_per_cluster`::
      Maximum number of connections maintained for the remote cluster when
      `sniff` mode is configured.
```

File: docs/reference/cluster/remote-info.asciidoc [hunk 9]
```diff
@@ -17,29 +17,47 @@ Returns configured remote cluster information.
...
+
+`address`::
+    The configured address for remote connections when proxy mode is configured.
```
Comment (jrodewig) [inline docs/reference/cluster/remote-info.asciidoc]:
```suggestion
    Address for remote connections when `proxy` mode is configured.
```

File: docs/reference/cluster/remote-info.asciidoc [hunk 10]
```diff
@@ -17,29 +17,47 @@ Returns configured remote cluster information.
...
+
+`num_sockets_connected`::
+    The number of open socket connections to the remote cluster when proxy mode
+    is configured.
```
Comment (jrodewig) [inline docs/reference/cluster/remote-info.asciidoc]:
```suggestion
    Number of open socket connections to the remote cluster when `proxy` mode
    is configured.
```

File: docs/reference/cluster/remote-info.asciidoc [hunk 11]
```diff
@@ -17,29 +17,47 @@ Returns configured remote cluster information.
...
+
+`max_socket_connections`::
+	The maximum number of socket connections to the remote cluster when proxy
+	mode is configured.
```
Comment (jrodewig) [inline docs/reference/cluster/remote-info.asciidoc]:
```suggestion
`max_socket_connections`::
    Maximum number of socket connections to the remote cluster when `proxy`
    mode is configured.
```

File: docs/reference/modules/remote-clusters.asciidoc [hunk 12]
```diff
@@ -15,21 +15,32 @@ endif::[]
 
 // tag::how-remote-clusters-work[]
 Remote cluster connections work by configuring a remote cluster and connecting
-only to a limited number of nodes in that remote cluster. Each remote cluster
-is referenced by a name and a list of seed nodes. When a remote cluster is
-registered, its cluster state is retrieved from one of the seed nodes and up
-to three _gateway nodes_ are selected to be connected to as part of remote
-cluster requests.
-// end::how-remote-clusters-work[]
+only to a limited number of nodes in that remote cluster. There are two
```
Comment (jrodewig) [inline docs/reference/modules/remote-clusters.asciidoc]:
```suggestion
to a limited number of nodes in that remote cluster. There are two
```

File: docs/reference/modules/remote-clusters.asciidoc [hunk 13]
```diff
@@ -15,21 +15,32 @@ endif::[]
...
 connections consist of uni-directional connections from the coordinating
-node to the selected remote _gateway nodes_ only.
+node to the remote remote connections.
+// end::how-remote-clusters-work[]
+
+[float]
+[[sniff-mode]]
+=== Sniff mode
+
+// tag::how-sniff-mode-works[]
```
Comment (jrodewig) [inline docs/reference/modules/remote-clusters.asciidoc]: It doesn't look like this section is reused anywhere. I would remove the tags until the content is actually reused.
```suggestion
```

File: docs/reference/modules/remote-clusters.asciidoc [hunk 14]
```diff
@@ -15,21 +15,32 @@ endif::[]
...
+A cluster in sniff mode is created by a name and a list of seed nodes. When a
+remote cluster is registered, its cluster state is retrieved from one of the
+seed nodes and up to three _gateway nodes_ are selected to be connected to as
+part of remote cluster requests. This mode requires that the _gateway node_
+publish addresses are exposed to the local cluster.
+
+The sniff mode is the default connection mode.
+// end::how-sniff-mode-works[]
```
Comment (jrodewig) [inline docs/reference/modules/remote-clusters.asciidoc]:
```suggestion
```

File: docs/reference/modules/remote-clusters.asciidoc [hunk 15]
```diff
@@ -15,21 +15,32 @@ endif::[]
...
-node to the selected remote _gateway nodes_ only.
+node to the remote remote connections.
+// end::how-remote-clusters-work[]
+
+[float]
+[[sniff-mode]]
+=== Sniff mode
+
+// tag::how-sniff-mode-works[]
+A cluster in sniff mode is created by a name and a list of seed nodes. When a
```
Comment (jrodewig) [inline docs/reference/modules/remote-clusters.asciidoc]:
```suggestion
A cluster in sniff mode is created using a name and a list of seed nodes. When a
```

File: docs/reference/modules/remote-clusters.asciidoc [hunk 16]
```diff
@@ -15,21 +15,32 @@ endif::[]
...
+remote cluster is registered, its cluster state is retrieved from one of the
+seed nodes and up to three _gateway nodes_ are selected to be connected to as
+part of remote cluster requests. This mode requires that the _gateway node_
```
Comment (jrodewig) [inline docs/reference/modules/remote-clusters.asciidoc]:
```suggestion
seed nodes and up to three _gateway nodes_ are selected as
part of remote cluster requests. This mode requires that the gateway node's
```

File: docs/reference/modules/remote-clusters.asciidoc [hunk 17]
```diff
@@ -62,9 +73,26 @@ communicate with 6.7. The matrix below summarizes compatibility as described abo
 (see <<remote-cluster-settings>>), though such tagged nodes still have
 to satisfy the two above requirements.
 
+[float]
+[[proxy-mode]]
+=== Proxy mode
+
+// tag::how-proxy-mode-works[]
+A cluster in proxy mode is created by a name and a single proxy address. When
```
Comment (jrodewig) [inline docs/reference/modules/remote-clusters.asciidoc]:
```suggestion
In proxy mode, a cluster is created using a name and a single proxy address. When
```

File: docs/reference/modules/remote-clusters.asciidoc [hunk 18]
```diff
@@ -62,9 +73,26 @@ communicate with 6.7. The matrix below summarizes compatibility as described abo
...
+a remote cluster is registered, a configurable number of socket connections are
+opened to the proxy address. The proxy is required to route those connections to
+the remote cluster. Proxy mode does not require that the remote cluster nodes
+publish addresses are accessible.
```
Comment (jrodewig) [inline docs/reference/modules/remote-clusters.asciidoc]:
```suggestion
the remote cluster. Proxy mode does not require remote cluster nodes
to have accessible publish addresses.
```

File: docs/reference/modules/remote-clusters.asciidoc [hunk 19]
```diff
@@ -62,9 +73,26 @@ communicate with 6.7. The matrix below summarizes compatibility as described abo
 (see <<remote-cluster-settings>>), though such tagged nodes still have
 to satisfy the two above requirements.
 
+[float]
+[[proxy-mode]]
+=== Proxy mode
+
+// tag::how-proxy-mode-works[]
```
Comment (jrodewig) [inline docs/reference/modules/remote-clusters.asciidoc]: I'd remove this unneeded tag.
```suggestion
```

File: docs/reference/modules/remote-clusters.asciidoc [hunk 20]
```diff
@@ -62,9 +73,26 @@ communicate with 6.7. The matrix below summarizes compatibility as described abo
...
+a remote cluster is registered, a configurable number of socket connections are
+opened to the proxy address. The proxy is required to route those connections to
+the remote cluster. Proxy mode does not require that the remote cluster nodes
+publish addresses are accessible.
+
+The proxy mode is not the default connection mode and must be configured. Similar
+to the sniff <<gateway-nodes-selection,gateway nodes>>, the remote
+connections are subject to the same version compatibility rules as
+<<rolling-upgrades>>.
+// end::how-proxy-mode-works[]
```
Comment (jrodewig) [inline docs/reference/modules/remote-clusters.asciidoc]:
```suggestion
```

File: docs/reference/modules/remote-clusters.asciidoc [hunk 21]
```diff
@@ -15,21 +15,32 @@ endif::[]
 
 // tag::how-remote-clusters-work[]
```
Comment (jrodewig) [inline docs/reference/modules/remote-clusters.asciidoc]: The content is this tagged region gets re-used here:
https://www.elastic.co/guide/en/elasticsearch/reference/master/modules-cross-cluster-search.html#ccs-works

Originally, that was intended to explain the concept of gateway nodes _before_ the recommendations outlined here:
https://www.elastic.co/guide/en/elasticsearch/reference/master/modules-cross-cluster-search.html#ccs-gateway-seed-nodes

With this changes, I don't think this re-use makes sense anymore. I'm also not sure if those recommendations would work with proxy mode.

Could you advise on how these changes would impact those CCS recommendations? After that, you can either update those docs as a part of this PR or I can update them as part of a separate PR. Thanks!

**Review by jrodewig — CHANGES_REQUESTED** [review]
I commented on several, non-blocking nits, but I do have a bigger concern with how this will impact the CCS docs here:

https://github.com/elastic/elasticsearch/pull/52779/files#r385049630

I think we should plan to address that either as part of this PR or a separate one I handle. Either way, it's not clear to me how this would impact CCS recommendations.

Thanks!

**Comment by jrodewig** [comment]
I took a stab at updating the CSS docs for remote cluster connection modes with e97d0216ce7062aedced50f47a1a7ae3929baece, which I pushed to this PR. This commits also removes some of the unneeded tags from the remote clusters docs.

That commit doesn't include some of the optional editorial suggestions I made. Feel free to take or leave those as you see fit. Also please feel free to change the CSS docs if needed.

**Review by jrodewig — COMMENTED** [review]
I took a stab at updating the CSS docs for remote cluster connection modes with e97d0216ce7062aedced50f47a1a7ae3929baece, which I pushed to this PR. This commit also removes some of the unneeded tags from the remote clusters docs.

That commit doesn't include some of the optional editorial suggestions I made. Feel free to take or leave those as you see fit. Also please feel free to change the CSS docs if needed.

**Comment by tbrooks8** [comment]
I updated this and added a section on proxy mode in the CCS docs.

File: docs/reference/modules/cross-cluster-search.asciidoc [hunk 22]
```diff
@@ -253,30 +253,33 @@ PUT _cluster/settings
...
-include::./remote-clusters.asciidoc[tag=how-remote-clusters-work]
-
 [discrete]
 [[ccs-gateway-seed-nodes]]
-=== Selecting gateway and seed nodes
+== Selecting gateway and seed nodes in sniff mode
 
-Gateway and seed nodes need to be accessible from the local cluster via your
-network.
+For remote clusters using the <<sniff-mode,sniff connection> mode, gateway and
```
Comment (jrodewig) [inline docs/reference/modules/cross-cluster-search.asciidoc]: This is required for the link to render properly.
```suggestion
For remote clusters using the <<sniff-mode,sniff connection>> mode, gateway and
```

File: docs/reference/modules/cross-cluster-search.asciidoc [hunk 23]
```diff
@@ -253,30 +253,33 @@ PUT _cluster/settings
...
 For {ccs}, we recommend you use gateway nodes that are capable of serving as
 <<coordinating-node,coordinating nodes>> for search requests. If
 wanted, the seed nodes for a cluster can be a subset of these gateway nodes.
 
+[discrete]
+[[ccs-proxy-mode]]
+== {ccs} in proxy mode
+
+<<proxy-mode,Proxy mode>> remote cluster connections support {ccs}. All remote
+connections will connect to the configured `proxy_address`. Any desired
```
Comment (jrodewig) [inline docs/reference/modules/cross-cluster-search.asciidoc]:
```suggestion
connections connect to the configured `proxy_address`. Any desired
```

File: docs/reference/modules/remote-clusters.asciidoc [hunk 24]
```diff
@@ -13,23 +13,31 @@ connections to a remote cluster. This functionality is used in
...
+
+[float]
+[[sniff-mode]]
+=== Sniff mode
+
+In sniff mode, a cluster is created using a name and a list of seed nodes. When
+a remote cluster is registered, its cluster state is retrieved from one of the
+seed nodes and up to three _gateway nodes_ are selected as part of remote
+cluster requests. This mode requires that the gateway node's publish addresses
+are exposed to the local cluster.
```
Comment (jrodewig) [inline docs/reference/modules/remote-clusters.asciidoc]: "accessible" is more consistent with what's stated in "proxy" mode and has less of a negative connotation.
```suggestion
cluster requests. This mode requires that the gateway node's publish addresses
are accessible by the local cluster.
```

File: docs/reference/modules/remote-clusters.asciidoc [hunk 25]
```diff
@@ -83,23 +106,32 @@ cluster:
...
-between local and remote indices.
+<1> `cluster_one`, `cluster_two`, and `cluster_three` are arbitrary _cluster aliases_
+representing the connection to each cluster. These names are subsequently used to
+distinguish between local and remote indices.
 <2> The hostname and <<modules-transport,transport>> port (default: 9300) of a
 seed node in the remote cluster.
 <3> A keep-alive ping is configured for `cluster_one`.
-<4> Compression is explicitly enabled for requests to `cluster_two`.
-<5> Disconnected remote clusters are optional for `cluster_two`.
+<4> The configured connection mode. By default it is <<sniff-mode,sniff>>, so
```
Comment (jrodewig) [inline docs/reference/modules/remote-clusters.asciidoc]:
```suggestion
<4> The configured connection mode. By default, this is <<sniff-mode,`sniff`>>, so
```

File: docs/reference/modules/remote-clusters.asciidoc [hunk 26]
```diff
@@ -251,6 +287,49 @@ PUT _cluster/settings
   Elasticsearch compresses the response. If unset, the global
   `transport.compress` is used as the fallback setting.
 
+[float]
+[[remote-cluster-sniff-settings]]
+=== Remote cluster settings for sniff mode
+
+`cluster.remote.${cluster_alias}.seeds`::
```
Comment (jrodewig) [inline docs/reference/modules/remote-clusters.asciidoc]: Our docs typically use the `<variable>` convention.
```suggestion
`cluster.remote.<cluster_alias>.seeds`::
```

File: docs/reference/modules/remote-clusters.asciidoc [hunk 27]
```diff
@@ -251,6 +287,49 @@ PUT _cluster/settings
...
+
+  The list of seed nodes used to sniff the remote cluster state.
+
+`cluster.remote.${cluster_alias}.node_connections`::
```
Comment (jrodewig) [inline docs/reference/modules/remote-clusters.asciidoc]:
```suggestion
`cluster.remote.<cluster_alias>.node_connections`::
```

File: docs/reference/modules/remote-clusters.asciidoc [hunk 28]
```diff
@@ -251,6 +287,49 @@ PUT _cluster/settings
...
+  A node attribute to filter out nodes that are eligible as a gateway node in
+  the remote cluster. For instance a node can have a node attribute
+  `node.attr.gateway: true` such that only nodes with this attribute will be
+  connected to if `cluster.remote.node.attr` is set to `gateway`.
+
+[float]
+[[remote-cluster-proxy-settings]]
+=== Remote cluster settings for proxy mode
+
+`cluster.remote.${cluster_alias}.proxy_address`::
```
Comment (jrodewig) [inline docs/reference/modules/remote-clusters.asciidoc]:
```suggestion
`cluster.remote.<cluster_alias>.proxy_address`::
```

File: docs/reference/modules/remote-clusters.asciidoc [hunk 29]
```diff
@@ -251,6 +287,49 @@ PUT _cluster/settings
...
+
+  The address used for all remote connections.
+
+`cluster.remote.${cluster_alias}.proxy_socket_connections`::
```
Comment (jrodewig) [inline docs/reference/modules/remote-clusters.asciidoc]:
```suggestion
`cluster.remote.<cluster_alias>.proxy_socket_connections`::
```

File: docs/reference/modules/remote-clusters.asciidoc [hunk 30]
```diff
@@ -251,6 +287,49 @@ PUT _cluster/settings
...
+
+  The number of socket connections to open per remote cluster. The default is
+  `18`.
+ifndef::include-xpack[]
+
+`cluster.remote.${cluster_alias}.server_name`::
```
Comment (jrodewig) [inline docs/reference/modules/remote-clusters.asciidoc]:
```suggestion
`cluster.remote.<cluster_alias>.server_name`::
```

Comment (jrodewig) [inline docs/reference/modules/remote-clusters.asciidoc]: I think this should be `ifdef`, not `ifndef`.

I'm also not sure if it's necessary to do `ifdef::include-xpack[]`. Adding `[role="xpack"]` before the setting def should suffice.

```suggestion

[role="xpack"]
`cluster.remote.${cluster_alias}.server_name`::
```

File: docs/reference/modules/remote-clusters.asciidoc [hunk 31]
```diff
@@ -251,6 +287,49 @@ PUT _cluster/settings
...
+
+  An optional hostname string which will be sent in the server_name field of
+  the TLS Server Name Indication extension if
+  <<configuring-tls,{TLS is enabled}>>. The TLS transport will fail to open
+  remote connections if this field is not a valid hostname as defined by the
+  TLS SNI specification.
+endif::[]
```
Comment (jrodewig) [inline docs/reference/modules/remote-clusters.asciidoc]:
```suggestion

```

File: docs/reference/modules/remote-clusters.asciidoc [hunk 32]
```diff
@@ -251,6 +287,49 @@ PUT _cluster/settings
...
+
+`cluster.remote.${cluster_alias}.proxy_socket_connections`::
+
+  The number of socket connections to open per remote cluster. The default is
+  `18`.
+ifndef::include-xpack[]
+
+`cluster.remote.${cluster_alias}.server_name`::
+
+  An optional hostname string which will be sent in the server_name field of
```
Comment (jrodewig) [inline docs/reference/modules/remote-clusters.asciidoc]:
```suggestion
  An optional hostname string which is sent in the server_name field of
```

File: docs/reference/modules/remote-clusters.asciidoc [hunk 33]
```diff
@@ -251,6 +287,49 @@ PUT _cluster/settings
...
+  the TLS Server Name Indication extension if
+  <<configuring-tls,{TLS is enabled}>>. The TLS transport will fail to open
```
Comment (jrodewig) [inline docs/reference/modules/remote-clusters.asciidoc]:
```suggestion
  <<configuring-tls,TLS is enabled>>. The TLS transport will fail to open
```

File: docs/reference/modules/remote-clusters.asciidoc [hunk 34]
```diff
@@ -139,7 +171,8 @@ PUT _cluster/settings
 // TEST[s/127.0.0.1:9300/\${transport_host}/]
 
 You can dynamically update the compression and ping schedule settings. However,
-you must re-include seeds in the settings update request. For example:
+you must re-include seeds or proxy_address in the settings update request.
```
Comment (jrodewig) [inline docs/reference/modules/remote-clusters.asciidoc]:
```suggestion
you must re-include seeds or a `proxy_address` in the settings update request.
```

File: docs/reference/modules/remote-clusters.asciidoc [hunk 35]
```diff
@@ -13,23 +13,31 @@ connections to a remote cluster. This functionality is used in
...
+[[sniff-mode]]
+=== Sniff mode
+
+In sniff mode, a cluster is created using a name and a list of seed nodes. When
+a remote cluster is registered, its cluster state is retrieved from one of the
+seed nodes and up to three _gateway nodes_ are selected as part of remote
+cluster requests. This mode requires that the gateway node's publish addresses
+are exposed to the local cluster.
+
+The sniff mode is the default connection mode.
```
Comment (jrodewig) [inline docs/reference/modules/remote-clusters.asciidoc]:
```suggestion
Sniff mode is the default connection mode.
```

**Review by jrodewig — APPROVED** [review]
Look good overall. I did suggest a few needed changes:

- Fix a broken internal link
- Remove or update the new `ifndef` directive in the remote cluster settings section
- Use the standard `<variable>` convention in the remote cluster settings section

My other feedback is non-blocking. Feel free to use or ignore those items as you see fit.

Thanks for writing up the proxy mode CCS section!

**Comment by tbrooks8** [comment]
I made @jrodewig changes. @ywelsch are you good with this?

**Review by ywelsch — APPROVED** [review]
LGTM

**Pull request merged**